db/config: Add SSTable compression options for user tables

ndragazis opened the pull request:
ScyllaDB offers the `compression` DDL property for configuring compression per user table (compression algorithm and chunk size). If not specified, the default compression algorithm is the LZ4Compressor with a 4KiB chunk size. The same default applies to system tables as well.

This series introduces a new configuration option to allow customizing the default for user tables. It also adds some tests for the new functionality.

Fixes #25195.

ndragazis committed:
db/config: Add SSTable compression options for user tables

ScyllaDB offers the `compression` DDL property for configuring
compression per user table (compression algorithm and chunk size). If
not specified, the default compression algorithm is the LZ4Compressor
with a 4KiB chunk size (refer to the default constructor for
`compression_parameters`). The same default applies to system tables as
well.

Add a new configuration option to allow customizing the default for user
tables. Use the previously hardcoded default as the new option's default
value.

Note that the option has no effect on ALTER TABLE statements. An altered
table either inherits explicit compression options from the CQL
statement, or maintains its existing options.

Signed-off-by: Nikos Dragazis <nikolaos.dragazis@scylladb.com>

mykaul commented:
> Additionally, disable compression for system tables by changing the default in the schema builder's `raw_schema` (no reason for them to use compression).

traces and audit are at least two exceptions I can think of for the above.

avikivity commented on db/config.hh:
```diff
@@ -431,6 +431,7 @@ public:
     named_value<bool> enable_sstables_mc_format;
     named_value<bool> enable_sstables_md_format;
     named_value<sstring> sstable_format;
+    named_value<string_map> sstable_compression_user_table_options;
```
We really should have the ability to write named_value\<my_struct>, but we don't.

xemul replied:
We do

```
named_value<seed_provider_type> seed_provider;
named_value<hinted_handoff_enabled_type> hinted_handoff_enabled;
named_value<std::vector<enum_option<experimental_features_t>>> experimental_features;
named_value<tri_mode_restriction> strict_allow_filtering;
named_value<std::vector<object_storage_endpoint_param>> object_storage_endpoints;
etc
```

ndragazis replied:
Resolved in v4.

avikivity commented on conf/scylla.yaml:
```diff
@@ -529,6 +529,28 @@ commitlog_total_space_in_mb: -1
+#   chunk_length_in_kb: 4
+#   crc_check_chance: 1.0
+
 # Relaxation of environment checks.
```
We generally don't add things to scylla.yaml, or it will grow without bounds.

mykaul replied:
But we do need to add it to our documentation.

nyh replied:
Have we documented somewhere what belongs in scylla.yaml?

Some projects add every single configuration option to their configuration file, which also doubles as documentation for the different options and their defaults. Clearly Scylla never went all-in with that approach - for example, I never listed **any** Alternator option in scylla.yaml (it seems recently @radoslawcybulski got on in - c36614e16d7c2c52b6921c0d6f9541c8acd21ef5).

Would be good to know what kind of options are supposed to be listed there and what options not to list.
If the existing scylla.yaml contains stuff it shouldn't, or missing stuff that it should contain, perhaps we should open an issue asking to fix that.

ndragazis replied:
> But we do need to add it to our documentation.

@mykaul our documentation (`configuration-parameters.rst`) fetches the config options directly from the code. So, the new option will appear in the docs regardless of scylla.yaml.

I checked what Cassandra is doing and I notice the same pattern as in Scylla - some options are absent from the yaml file. But I couldn't find any reference about what options are excluded.

ndragazis commented on cql3/statements/cf_prop_defs.cc:
```diff
@@ -353,9 +354,14 @@ void cf_prop_defs::apply_to_builder(schema_builder& builder, schema::extensions_
         builder.set_compressor_params(compression_parameters(*compression_options));
+    } else if (!is_system_keyspace(ks_name)) {
+        builder.set_compressor_params(compression_parameters(db.get_config().sstable_compression_user_table_options() | std::ranges::to<std::map>()));
     }
```
Hmm, this would indeed override the existing compression settings (regardless of their values) with the default. Thanks for pointing this out, I missed it.

```
cqlsh:ks> CREATE TABLE t (pk int PRIMARY KEY) with compression= {'sstable_compression': ''};
cqlsh:ks> describe t

CREATE TABLE ks.t (
    pk int,
    PRIMARY KEY (pk)
) WITH bloom_filter_fp_chance = 0.01
    AND caching = {'keys': 'ALL', 'rows_per_partition': 'ALL'}
    AND comment = ''
    AND compaction = {'class': 'IncrementalCompactionStrategy'}
    AND compression = {}
    AND crc_check_chance = 1
    AND default_time_to_live = 0
    AND gc_grace_seconds = 864000
    AND max_index_interval = 2048
    AND memtable_flush_period_in_ms = 0
    AND min_index_interval = 128
    AND speculative_retry = '99.0PERCENTILE'
    AND tombstone_gc = {'mode': 'timeout', 'propagation_delay_in_seconds': '3600'};

cqlsh:ks> ALTER TABLE t WITH comment = 'foo';
cqlsh:ks> describe t

CREATE TABLE ks.t (
    pk int,
    PRIMARY KEY (pk)
) WITH bloom_filter_fp_chance = 0.01
    AND caching = {'keys': 'ALL', 'rows_per_partition': 'ALL'}
    AND comment = 'foo'
    AND compaction = {'class': 'IncrementalCompactionStrategy'}
    AND compression = {'chunk_length_in_kb': '4', 'crc_check_chance': '1.000000', 'sstable_compression': 'org.apache.cassandra.io.compress.LZ4Compressor'}
    AND crc_check_chance = 1
    AND default_time_to_live = 0
    AND gc_grace_seconds = 864000
    AND max_index_interval = 2048
    AND memtable_flush_period_in_ms = 0
    AND min_index_interval = 128
    AND speculative_retry = '99.0PERCENTILE'
    AND tombstone_gc = {'mode': 'timeout', 'propagation_delay_in_seconds': '3600'};
```

bhalevy replied:
We discussed a similar scenario regarding the tablets options.
IMO we should document that unless specified otherwise option groups that are omitted from ALTER TABLE retain their current value.  As for the semantics of each option group and its sub-optiions, this would probably be option-group specific.
Again, unless otherwise specified we can say that the option-group is set as a whole by ALTER TABLE overriding all the group's sub-options.

avikivity commented:
/cc @xemul who knows the table properties better

ndragazis commented on cql3/statements/create_table_statement.cc:
```diff
@@ -122,6 +123,14 @@ void create_table_statement::apply_properties_to(schema_builder& builder, const
 #endif
 
+    // Set up default compression for user tables.
+    if (!_properties->get_compression_options()) {
```
That's where I put it in v1, but this breaks ALTER TABLE which also calls `apply_to_builder`, see: https://github.com/scylladb/scylladb/pull/26003#discussion_r2355043727

ndragazis commented on cql3/statements/create_table_statement.cc:
```diff
@@ -122,6 +123,14 @@ void create_table_statement::apply_properties_to(schema_builder& builder, const
 
+    // Set up default compression for user tables.
+    if (!_properties->get_compression_options()) {
+        if (is_system_keyspace(keyspace())) {
```
Yes, it is being checked by `create_table_statement::check_access()`. This check here serves as an assertion (didn't use `SCYLLA_ASSERT` per https://github.com/scylladb/scylladb/issues/7871).

The problem is that it breaks this test: `cqlpy.test_tools.TestScyllaSsstableSchemaLoading.test_external_dir_schema_file`. This test runs `sstable dump-data` on an SSTable from `system.scylla_local`, which internally calls `create_table_statement::get_cf_meta_data()` to build the schema without actually creating the table, so there is no prior call to `check_access()`. Yet another corner case I missed.

ndragazis committed:
test/boost: Add tests for SSTable compression config options

Since patch 03461d6a54, all boost unit tests depending on `cql_test_env`
are compiled into a single executable (`combined_tests`). Add the new
test in there.

Signed-off-by: Nikos Dragazis <nikolaos.dragazis@scylladb.com>

ndragazis committed:
db/config: Prepare compression_parameters for config system

SSTable compression is currently configurable only per table, via the
`compression` property in CREATE/ALTER TABLE statements. This is
represented internally via the `compression_parameters` class. We plan
to offer the same options via the configuration as well, to make the
default compression method for user tables configurable.

This patch prepares the ground by making the `compression_parameters`
usable as a `config_file::named_value`, namely:

* Define an extraction operator (required by `boost::program_options`
  for parsing the options from command line).
* Define a formatter (required by `named_value::operator()`).
* Define a template specialization for `config_type_for` (required by
  `named_value` constructor).
* Define a yaml converter (required for parsing the options from
  scylla.yaml).

Signed-off-by: Nikos Dragazis <nikolaos.dragazis@scylladb.com>

ndragazis committed:
main: Validate SSTable compression options from config

`compression_parameters` provides two levels of validation:

* syntactic checks - implemented in the constructor
* semantic checks - implemented by `compression_parameters::validate()`

The former are applied implicitly when parsing the options from the
command line or from scylla.yaml. The latter are currently not applied,
but they should.

In lack of a better place, apply them in main, right after joining the
cluster, to make sure that the cluster features have been negotiated.
The feature needed here is the `SSTABLE_COMPRESSION_DICTS`. Validation
will fail if the feature is disabled and a dictionary compression
algorithm has been selected.

Also, mark `validate()` as const so that it can be called from a config
object.

Signed-off-by: Nikos Dragazis <nikolaos.dragazis@scylladb.com>

ndragazis committed:
test/cluster: Add tests for invalid SSTable compression options

Complementary to the previous patch. It triggers semantic validation
checks in `compression_parameters::validate()` and expects the server to
exit. The tests examine both command line and YAML options.

Signed-off-by: Nikos Dragazis <nikolaos.dragazis@scylladb.com>

ndragazis committed:
compressor: Add missing space in exception message

Signed-off-by: Nikos Dragazis <nikolaos.dragazis@scylladb.com>

ndragazis committed:
compressor: Validate presence of sstable_compression in parameters

SSTable compression parameters should always define an algorithm via the
`sstable_compression` sub-option. Add a check in the constructor to
ensure this is always provided (unless no options are given, which is
interpreted as "no compression").

This change has no user-visible effect, since the same check is already
performed at a higher-level, while validating the CQL properties of
CREATE TABLE and ALTER TABLE statements (see `cf_prop_defs::validate()`).
However, it will become useful in later patches, when compression config
options will be introduced.

Although now redundant, keep the sanity check in
`cf_prop_defs::validate()` to maintain consistency of error messages
with Cassandra.

Note also that Cassandra uses 'class' instead of 'sstable_compression'
since version 3.11.10, but Scylla still doesn't support this, see:
https://github.com/scylladb/scylladb/issues/4200

Signed-off-by: Nikos Dragazis <nikolaos.dragazis@scylladb.com>

ndragazis commented:
> #### Failed Tests (9/78655):
> 
>  * [test_describe](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/20394/testReport/junit/cqlsh_tests.cqlsh_tests/TestCqlsh/release___dtest_with_consistent_topology_changes___test_describe) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_describe)
>  * [test_describe](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/20394/testReport/junit/cqlsh_tests.cqlsh_tests/TestCqlsh/release___dtest_with_tablets___test_describe) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_describe)
>  * [test_describe](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/20394/testReport/junit/cqlsh_tests.cqlsh_tests/TestCqlsh/release___dtest_with_gossip_topology_changes___test_describe) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_describe)
>  * [test_describe[require_client_auth=false]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/20394/testReport/junit/cqlsh_tests.cqlsh_tests/TestCqlshWithSSL/release___dtest_with_gossip_topology_changes___test_describe_require_client_auth_false_) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_describe%5Brequire_client_auth=false%5D)
>  * [test_describe[require_client_auth=false]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/20394/testReport/junit/cqlsh_tests.cqlsh_tests/TestCqlshWithSSL/release___dtest_with_consistent_topology_changes___test_describe_require_client_auth_false_) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_describe%5Brequire_client_auth=false%5D)
>  * [test_describe[require_client_auth=false]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/20394/testReport/junit/cqlsh_tests.cqlsh_tests/TestCqlshWithSSL/release___dtest_with_tablets___test_describe_require_client_auth_false_) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_describe%5Brequire_client_auth=false%5D)
>  * [test_describe[require_client_auth=true]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/20394/testReport/junit/cqlsh_tests.cqlsh_tests/TestCqlshWithSSL/release___dtest_with_gossip_topology_changes___test_describe_require_client_auth_true_) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_describe%5Brequire_client_auth=true%5D)
>  * [test_describe[require_client_auth=true]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/20394/testReport/junit/cqlsh_tests.cqlsh_tests/TestCqlshWithSSL/release___dtest_with_tablets___test_describe_require_client_auth_true_) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_describe%5Brequire_client_auth=true%5D)
>  * [test_describe[require_client_auth=true]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/20394/testReport/junit/cqlsh_tests.cqlsh_tests/TestCqlshWithSSL/release___dtest_with_consistent_topology_changes___test_describe_require_client_auth_true_) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_describe%5Brequire_client_auth=true%5D)

The tests fail because the compression property from `DESCRIBE TABLE` contains a `chunk_length_in_kb` and `crc_check_chance`:
```
AND compression={'chunk_length_in_kb':'4','crc_check_chance':'1.000000','sstable_compression':'org.apache.cassandra.io.compress.LZ4Compressor'}
```
The expected value is:
```
AND compression = {{'sstable_compression': 'org.apache.cassandra.io.compress.LZ4Compressor'}}
```
It turns out that if you don't explicitly define these properties, they do not appear in the table schema. So, I should better not provide explicit defaults for the new config option.

ndragazis commented:
v5 Changelog:

```diff
-    , sstable_compression_user_table_options(this, "sstable_compression_user_table_options", value_status::Used,
-        std::map<sstring, sstring>{
-            {compression_parameters::SSTABLE_COMPRESSION, "LZ4Compressor"},
-            {compression_parameters::CHUNK_LENGTH_KB, "4"},
-            {compression_parameters::CRC_CHECK_CHANCE, "1.0"}
-        },
+    , sstable_compression_user_table_options(this, "sstable_compression_user_table_options", value_status::Used, compression_parameters{},
```

michoecho commented:
> Also, but less severe - I'm not sure why we chose level 3 for zstd. Based on what?

@mykaul It's zstd's default level. And, because of that, it has also been Scylla's default level since forever. It's the most conservative choice. Picking something else is what would require a rationale.

mykaul commented:
> > Also, but less severe - I'm not sure why we chose level 3 for zstd. Based on what?
> 
> @mykaul It's zstd's default level. And, because of that, it has also been Scylla's default level since forever. It's the most conservative choice. Picking something else is what would require a rationale.

I'm looking at some random results I got. Example:
```json
{"level": 1, "chunk_length_in_kb": 4, "dict": "future", "sstable_compression": "ZstdWithDictsCompressor", "ratio": 0.311159},
{"level": 3, "chunk_length_in_kb": 4, "dict": "future", "sstable_compression": "ZstdWithDictsCompressor", "ratio": 0.30916625},
```
For 0.2% better compression, how much CPU usage difference there is between the levels?

(this is just one sample, need to look at other samples too).

For reference to the above, same data set:
```json
{"level": 1, "chunk_length_in_kb": 4, "dict": "future", "sstable_compression": "LZ4WithDictsCompressor", "ratio": 0.40414393},
```
so zstd is saving additional ~10%, nice.

and:
```json
{"level": 5, "chunk_length_in_kb": 16, "dict": "future", "sstable_compression": "ZstdWithDictsCompressor", "ratio": 0.29073685}
```

And increasing both the level and chunk level to 5 and 16, doesn't gain us much over the above. Just additional ~2%.
